linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -1518,19 +1518,53 @@ private class FunctionEmitter private (
     val BinaryOp(op, lhs, rhs) = tree
     assert(op == === || op == !==)
 
-    // TODO Optimize this when the operands have a better type than `any`
-
-    genTree(lhs, AnyType)
-    genTree(rhs, AnyType)
+    def maybeGenInvert(): Unit = {
+      if (op == BinaryOp.!==)
+        fb += wa.I32Eqz
+    }
 
-    markPosition(tree)
+    val lhsType = lhs.tpe
+    val rhsType = rhs.tpe
 
-    fb += wa.Call(genFunctionID.is)
+    if (lhsType == NothingType) {
+      genTree(lhs, NothingType)
+      NothingType
+    } else if (rhsType == NothingType) {
+      genTree(lhs, NoType)
+      genTree(rhs, NothingType)
+      NothingType
+    } else if (rhsType == NullType) {
+      /* Note that the optimizer normalizes Literals on the right of `===`,
+       * so testing for the `lhsType == NullType` is not as useful.
+       */
+      genTree(lhs, AnyType)
+      genTree(rhs, NoType) // no-op if it is actually a Null() literal
+      markPosition(tree)
+      fb += wa.RefIsNull
+      maybeGenInvert()
+      BooleanType
+    } else {
```

Comment:
IMO this is the optimizer's job, indeed.

It would be a nice-to-have to do it, but it's not that critical. Usually, comparisons between primitives are already produced as `Int_==` (for example) by the compiler.